internal/app/singularity/remote_add_test.go

```diff
@@ -0,0 +1,300 @@
+			},
+		},
+	}
+
+	yaml, err := yaml.Marshal(cfg)
+	if err != nil {
+		f.Close()
+		os.Remove(path)
+		t.Fatalf("cannot marshal YAML: %s\n", err)
+	}
+
+	f.Write(yaml)
+	f.Close()
+
+	return path
+}
+
+func TestRemoteAddAndRemove(t *testing.T) {
+	validCfgFile := createValidCfgFile(t)
+	defer os.Remove(validCfgFile)
+
+	invalidCfgFile := createInvalidCfgFile(t)
+	defer os.Remove(invalidCfgFile)
+
+	tests := []struct {
+		name       string
+		cfgfile    string
+		remoteName string
+		uri        string
+		global     bool
+		shallPass  bool
+	}{
+		{
+			name:       "1: invalid config file; empty remote name; invalid URI, local",
+			cfgfile:    invalidCfgFile,
+			remoteName: "",
+			uri:        invalidURI,
+			global:     false,
+			shallPass:  true,
+		},
+		{
+			name:       "2: invalid config file; empty remote name; empty URI; local",
+			cfgfile:    invalidCfgFile,
+			remoteName: "",
+			uri:        "",
+			global:     false,
+			shallPass:  true,
+		},
+		{
+			name:       "3: invalid config file; invalid remote name; invalid URI; local",
+			cfgfile:    invalidCfgFile,
+			remoteName: invalidRemoteName,
+			uri:        invalidURI,
+			global:     false,
+			shallPass:  true,
+		},
+		{
+			name:       "4: invalid config file; invalid remote name; invalid URI; local",
+			cfgfile:    invalidCfgFile,
+			remoteName: invalidRemoteName,
+			uri:        invalidURI,
+			global:     false,
+			shallPass:  true,
+		},
+		{
+			name:       "5: valid config file; empty remote name; invalid URI, local",
+			cfgfile:    validCfgFile,
+			remoteName: "",
+			uri:        invalidURI,
+			global:     false,
+			shallPass:  true,
+		},
+		{
+			name:       "6: valid config file; empty remote name; empty URI; local",
+			cfgfile:    validCfgFile,
+			remoteName: "",
+			uri:        "",
+			global:     false,
+			shallPass:  true,
+		},
+		{
+			name:       "7: valid config file; invalid remote name; invalid URI; local",
+			cfgfile:    validCfgFile,
+			remoteName: invalidRemoteName,
+			uri:        invalidURI,
+			global:     false,
+			shallPass:  true,
+		},
+		{
+			name:       "8: valid config gile; invalid remote name; invalid URI; local",
+			cfgfile:    validCfgFile,
+			remoteName: invalidRemoteName,
+			uri:        invalidURI,
+			global:     false,
+			shallPass:  true,
+		},
+		{
+			// This test checks both RemoteAdd() and RemoteRemove(), we stil
+			// have a separate test for corner cases in the context of
+			// RemoveRemove().
+			name:       "9: valid config file; valid remote name; valid URI; local",
+			cfgfile:    validCfgFile,
+			remoteName: "cloud_testing",
+			uri:        "cloud.random.io",
+			global:     false,
+			shallPass:  true,
+		},
+		{
+			name:       "10: valid but dummy file; empty remote name; empty URI; local",
```

Comment:
I think this is a duplicate of 2?

Reply:
I made a pass over all the cases, removed the duplicates and added the missing cases.